tests/client/handshake.rs

```diff
@@ -0,0 +1,50 @@
+use bson::{self, Bson};
+use mongodb::{DRIVER_NAME, Client, ThreadedClient};
+use mongodb::db::ThreadedDatabase;
+use mongodb::CommandType;
+
+#[derive(Debug, Deserialize)]
+struct Metadata {
+    #[serde(rename = "clientMetadata")]
+    pub client: ClientMetadata,
+}
+
+#[derive(Debug, Deserialize)]
+struct ClientMetadata {
+    pub driver: DriverMetadata,
+    pub os: OsMetadata,
+}
+
+#[derive(Debug, Deserialize)]
+struct DriverMetadata {
+    pub name: String,
+    pub version: String,
+}
+
+#[derive(Debug, Deserialize)]
+struct OsMetadata {
+    #[serde(rename = "type")]
+    pub os_type: String,
+    pub architecture: String,
+}
+
+#[test]
+fn metadata_sent_in_handshake() {
+    let client = Client::connect("localhost", 27017).unwrap();
+    let db = client.db("admin");
+    let version = db.version().unwrap();
+
+    if version.major < 3 || (version.major == 3 && version.minor < 4) {
```

Comment:
The server only recognizes client metadata starting in 3.4. We don't bother checking the version within the driver because the server will silently ignore the metadata in older versions, but the server won't give back the metadata in those versions, so we skip the assertions.

Reply:
Are there any other options that check that the version is less than 3.4?  If so, it would make sense to make a function for this.

Reply:
We currently don't have any other tests that only run above a certain version (since most 3.4+ features are not implemented yet), but I like the idea of abstracting this out for future work.